src/CodeGen_AMDGPU_Dev.cpp

```diff
@@ -0,0 +1,507 @@
+        sym_push(loop->name, codegen(simt_idx));
+        codegen(loop->body);
+        sym_pop(loop->name);
+    } else {
+        CodeGen_LLVM::visit(loop);
+    }
+}
+
+void CodeGen_AMDGPU_Dev::visit(const Allocate *alloc) {
+    user_assert(!alloc->new_expr.defined()) << "Allocate node inside AMDGPU kernel has custom new expression.\n" <<
+        "(Memoization is not supported inside GPU kernels at present.)\n";
+    if (alloc->name == "__shared") {
+        // AMDGPU uses zero in address space 3 as the base address for shared memory
+        Value *shared_base = Constant::getNullValue(PointerType::get(i8_t, 3));
+        sym_push(alloc->name, shared_base);
+    } else {
+        debug(2) << "Allocate " << alloc->name << " on device\n";
+
+        string allocation_name = alloc->name;
+        debug(3) << "Pushing allocation called " << allocation_name << " onto the symbol table\n";
+
+        // Jump back to the entry and generate an alloca. Note that by
+        // jumping back we're rendering any expression we carry back
+        // meaningless, so we had better only be dealing with
+        // constants here.
+        int32_t size = alloc->constant_allocation_size();
+        user_assert(size > 0)
+            << "Allocation " << alloc->name << " has a dynamic size. "
+            << "Only fixed-size allocations are supported on the gpu. "
+            << "Try storing into shared memory instead.";
+
+        BasicBlock *here = builder->GetInsertBlock();
+
+        builder->SetInsertPoint(entry_block);
+        Value *ptr = builder->CreateAlloca(llvm_type_of(alloc->type), ConstantInt::get(i32_t, size));
+        builder->SetInsertPoint(here);
+        sym_push(allocation_name, ptr);
+    }
+    codegen(alloc->body);
+}
+
+void CodeGen_AMDGPU_Dev::visit(const Free *f) {
+    sym_pop(f->name);
+}
+
+void CodeGen_AMDGPU_Dev::visit(const AssertStmt *op) {
+    // Discard the error message for now.
+    Expr trap = Call::make(Int(32), "halide_ptx_trap", {}, Call::Extern);
+    codegen(IfThenElse::make(!op->condition, Evaluate::make(trap)));
+}
+
+void CodeGen_AMDGPU_Dev::visit(const Load *op) {
+
+    // Do aligned 4-wide 32-bit loads as a single i128 load.
+    const Ramp *r = op->index.as<Ramp>();
+    // TODO: lanes >= 4, not lanes == 4
+    if (is_one(op->predicate) && r && is_one(r->stride) && r->lanes == 4 && op->type.bits() == 32) {
+        ModulusRemainder align = modulus_remainder(r->base, alignment_info);
+        if (align.modulus % 4 == 0 && align.remainder % 4 == 0) {
+            Expr index = simplify(r->base / 4);
+            Expr equiv = Load::make(UInt(128), op->name, index,
+                                    op->image, op->param, const_true());
+            equiv = reinterpret(op->type, equiv);
+            codegen(equiv);
+            return;
+        }
+    }
+
+    CodeGen_LLVM::visit(op);
+}
+
+void CodeGen_AMDGPU_Dev::visit(const Store *op) {
+
+    // Do aligned 4-wide 32-bit stores as a single i128 store.
```

Comment:
This Store visitor was to work around a particular piece of weirdness in the nvptx backend. Not necessary here.

Reply:
Ok. So, how should I implement the Store visitor?

Reply:
I think the default store visitor will work fine. Try just deleting this method.

Reply:
Do you want me to get rid of the whole `visit(const Store *op)` implementation?